### ISSUE TITLE
ERROR TypeError: useEditor is not a constructor on loading data from api

### ISSUE THREAD

**Issue by sreenithalbhat** [ghiscoding/Angular-Slickgrid#195]
### Your Environment

| Software          | Version(s) |
| ----------------- | ---------- |
| Angular           |^6.0.3
| Angular-Slickgrid |  ^1.9.6|
| TypeScript        | ~2.7.2
| Operating System  |
| NPM/Node/Yarn     |

### Context
When enabling edition in the grid's options and setting editors for each column definitions, I get the following error when I try to edit a cell:
```
core.js:1598 ERROR TypeError: useEditor is not a constructor
    at makeActiveCellEditable (slick.grid.js:4410)
    at setActiveCellInternal (slick.grid.js:4297)
    at gotoCell (slick.grid.js:5064)
    at HTMLDivElement.handleDblClick (slick.grid.js:4040)
    at HTMLDivElement.dispatch (jquery-3.2.1.slim.min.js:3)
    at HTMLDivElement.push../node_modules/slickgrid/lib/jquery.event.drag-2.3.0.js.$event.dispatch (jquery.event.drag-2.3.0.js:382)
    at HTMLDivElement.q.handle (jquery-3.2.1.slim.min.js:3)
    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:421)
    at Object.onInvokeTask (core.js:4053)
    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:420)
defaultErrorLogger @ core.js:1598
push../node_modules/@angular/core/fesm5/core.js.ErrorHandler.handleError @ core.js:1647
next @ core.js:4727
schedulerFn @ core.js:3712
push../node_modules/rxjs/_esm5/internal/Subscriber.js.SafeSubscriber.__tryOrUnsub @ Subscriber.js:196
push../node_modules/rxjs/_esm5/internal/Subscriber.js.SafeSubscriber.next @ Subscriber.js:134
push../node_modules/rxjs/_esm5/internal/Subscriber.js.Subscriber._next @ Subscriber.js:77
push../node_modules/rxjs/_esm5/internal/Subscriber.js.Subscriber.next @ Subscriber.js:54
push../node_modules/rxjs/_esm5/internal/Subject.js.Subject.next @ Subject.js:47
push../node_modules/@angular/core/fesm5/core.js.EventEmitter.emit @ core.js:3704
(anonymous) @ core.js:4084
push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke @ zone.js:388
push../node_modules/zone.js/dist/zone.js.Zone.run @ zone.js:138
push../node_modules/@angular/core/fesm5/core.js.NgZone.runOutsideAngular @ core.js:4021
onHandleError @ core.js:4084
push../node_modules/zone.js/dist/zone.js.ZoneDelegate.handleError @ zone.js:392
push../node_modules/zone.js/dist/zone.js.Zone.runTask @ zone.js:191
push../node_modules/zone.js/dist/zone.js.ZoneTask.invokeTask @ zone.js:496
invokeTask @ zone.js:1540
globalZoneAwareCallback @ zone.js:1566
```

### Possible Solution
i tried `this.columnDefinitions = this.columnDefinitions.slice();` ..but not working

### Code Sample
```ts
this.columnDefinitions.push(retVal.header); //retVal.header is the columndefinition array from api
this.columnDefinitions = this.columnDefinitions.slice();
          this.formattedData = retVal.data;
          
          this.gridObj.invalidate();
          this.gridObj.render();
```

**Comment by ghiscoding** [comment]
There not enough information for me to help you, but it seems that you are trying to add a column dynamically. There's an example [here](https://ghiscoding.github.io/Angular-Slickgrid/#/localization) and it's Component [here](https://github.com/ghiscoding/Angular-Slickgrid/blob/master/src/app/examples/grid-localization.component.ts#L151)

**Comment by sreenithalbhat** [comment]
I am trying to hide and show certain column on changing the tab.But while i change the tab columns are hidden/shown but not able to edit an am getting the above error.

this.visibleColumns = this.columnDefinitions.slice(0, 1); this.visibleColumns.push(...this.columnDefinitions.slice(13, 25));

**Comment by ghiscoding** [comment]
Without a real sample, it's hard to help you but if you just want to show/hide certain columns (which is a lot easier than adding/removing), you can look at the [Grid Presets](https://github.com/ghiscoding/Angular-Slickgrid/wiki/Grid-State-&-Preset#example). If you want to do something similar to the preset but not just at the start but also at any point in time, then it seems to be only 1 line of code
```ts
grid.setColumns(updatedColumns);
```
How it works is simple, it is the full list of column definitions, whatever is not in the list when you call `setColumns` will be hidden. Also the order of that list represent the order of each column in the grid. If you then want to show the column(s) again, then redo the exercise of calling `setColumns()`.

**Comment by reejeshkv** [comment]
I too want to display/ hide columns dynamically. I tried slice() and looks like it is working but the issue is whenever there is an 'editor' in cells those are not working after hide/show the columns.

I am getting "ERROR TypeError: useEditor is not a constructor"  error while clicking those dynamically added cells with editor.

**Comment by ghiscoding** [comment]
Like I said earlier, show/hide is not the same as add/remove columns. Show/hide is a lot easier to do and you don't need `slice()` for that. Adding/removing dynamically might have some issues with the editor not being created properly. The Editor might not work because I use a different property name in Angular-Slickgrid, I use the `internalColumnEditor`, you can see some details on this [line](https://github.com/ghiscoding/Angular-Slickgrid/blob/master/src/app/modules/angular-slickgrid/components/angular-slickgrid.component.ts#L246) of Angular-Slickgrid internal code. For more info read below. When you create new column dynamically, you will have to do that job manually of copy the editor into the correct property. Again see the [line](https://github.com/ghiscoding/Angular-Slickgrid/blob/master/src/app/modules/angular-slickgrid/components/angular-slickgrid.component.ts#L246) to understand what to do

Internally I take the `editor` and copy it to `internalColumnEditor` and then copy `editor.model` to `editor`. Why do I do that? Because SlickGrid has it's own Editor factory which doesn't work well with dome DI (like ngx-translate) that I need for the Editors and also there's no Filter factory. I wanted both the Editor & Filter to be of the same structure (in your code) to make it easier to access when creating a grid.

**Comment by ghiscoding** [comment]
I made a change in the library and pushed a new version `2.7.2` but it seems to still be a problem. Not sure why though since it was working locally in the lib, I will have to recheck it.

**Comment by ghiscoding** [comment]
It's fixed and now working with version `2.7.2`, you can see it working in the [Demo](https://ghiscoding.github.io/Angular-Slickgrid/#/editor)

If you like the lib, please up vote if you haven't yet ⭐️ 
Cheers

**Comment by sreenithalbhat** [comment]
> Like I said earlier, show/hide is not the same as add/remove columns. Show/hide is a lot easier to do and you don't need `slice()` for that.
> 
> If you still want to do adding/removing dynamically, I now understand your problem, it might have to do with the code that I've put in place within Angular-Slickgrid to deal with Editors. The Editor might not work because I use a different property name in Angular-Slickgrid, I use the `internalColumnEditor`, you can see some details on this [line](https://github.com/ghiscoding/Angular-Slickgrid/blob/master/src/app/modules/angular-slickgrid/components/angular-slickgrid.component.ts#L246) of Angular-Slickgrid internal code. For more info read below. When you create new column dynamically, you will have to do that job manually of copy the editor into the correct property. Again see the [line](https://github.com/ghiscoding/Angular-Slickgrid/blob/master/src/app/modules/angular-slickgrid/components/angular-slickgrid.component.ts#L246) to understand what to do.
> 
> Internally I take the `editor` and copy it to `internalColumnEditor` and then copy `editor.model` to `editor`. Why do I do that? Because SlickGrid has it's own Editor factory which doesn't work well with dome DI (like ngx-translate) that I need for the Editors and also there's no Filter factory. I wanted both the Editor & Filter to be of the same structure (in your code) to make it easier to access when creating a grid.
> 
> I am not responsible for the Editor dynamic creation, you'll have to read what I wrote and do the same on your side.

this worked for me.Thanks you.Great work!
